File: packages/core/go/protocol.go
package botiva

import (
	"encoding/json"
	"strings"
	"time"
)

// ProtocolVersion of the Botiva Wire Protocol this port speaks.
const ProtocolVersion = "botiva/1"

// Frame is one wire frame. A map keeps the JSON mapping transparent and the
// port compact; typed accessors live where they matter (seq, type).
type Frame = map[string]any

// IncomingMessage is a parsed user message.
type IncomingMessage struct {
	Text string
	ID   string
	Meta map[string]any
}

// HelloFrame is the client handshake.
type HelloFrame struct {
	UserID         string
	ConversationID string
	Watermark      int
	HasWatermark   bool
	Token          string // auth credential (§2.1), when the server authenticates
	Meta           map[string]any
}

// Inbound is the result of ParseIncoming: exactly one field is non-nil.
type Inbound struct {
	Hello   *HelloFrame
	Message *IncomingMessage
}

// ParseIncoming accepts a JSON string, raw bytes, a parsed frame map or plain
// text — mirrors parseIncoming in @botiva/core.
func ParseIncoming(raw any) *Inbound {
	var value map[string]any
	switch v := raw.(type) {
	case string:
		if err := json.Unmarshal([]byte(v), &value); err != nil {
			text := strings.TrimSpace(v)
			if text == "" {
				return nil
			}
			return &Inbound{Message: &IncomingMessage{Text: text}}
		}
	case []byte:
		return ParseIncoming(string(v))
	case map[string]any:
		value = v
	default:
		return nil
	}
	if value == nil {
		return nil
	}
	if value["type"] == "hello" {
		hello := &HelloFrame{}
		hello.UserID, _ = value["userId"].(string)
		hello.ConversationID, _ = value["conversationId"].(string)
		if w, ok := value["watermark"].(float64); ok {
			hello.Watermark = int(w)
			hello.HasWatermark = true
		}
		hello.Token, _ = value["token"].(string)
		hello.Meta, _ = value["meta"].(map[string]any)
		return &Inbound{Hello: hello}
	}
	text := ""
	if data, ok := value["data"].(map[string]any); ok {
		text, _ = data["text"].(string)
	}
	if text == "" {
		text, _ = value["text"].(string)
	}
	text = strings.TrimSpace(text)
	if text == "" {
		return nil
	}
	msg := &IncomingMessage{Text: text}
	msg.ID, _ = value["id"].(string)
	msg.Meta, _ = value["meta"].(map[string]any)
	return &Inbound{Message: msg}
}

// FrameMapping pairs a frame with its persistence class.
type FrameMapping struct {
	Frame      Frame
	Persistent bool
}

// EventToFrames is the canonical AgentEvent → wire frame mapping.
// Must stay byte-compatible with @botiva/core eventToFrames (PROTOCOL.md §4).
func EventToFrames(ev AgentEvent, newID func(prefix string) string) []FrameMapping {
	now := time.Now().UnixMilli()
	textFrame := func(text string, extra Frame) Frame {
		f := Frame{
			"type":      "text",
			"id":        newID("msg"),
			"from":      "bot",
			"data":      map[string]any{"text": text},
			"timestamp": now,
		}
		for k, v := range extra {
			f[k] = v
		}
		return f
	}

	switch ev.Type {
	case "message":
		extra := Frame{}
		if len(ev.Actions) > 0 {
			extra["actions"] = ev.Actions
		}
		return []FrameMapping{{textFrame(ev.Text, extra), true}}
	case "tool_call":
		return []FrameMapping{{Frame{"type": "tool_call", "data": ev.ToolCall}, true}}
	case "genui":
		streamID := ev.StreamID
		if streamID == "" {
			streamID = newID("stream")
		}
		return []FrameMapping{{Frame{
			"type": "genui", "streamId": streamID, "chunk": ev.Chunk, "done": ev.Done,
		}, true}}
	case "interrupt":
		question := "Your confirmation is needed to continue."
		options := []any{"Approve", "Cancel"}
		switch p := ev.Payload.(type) {
		case string:
			question = p
		case map[string]any:
			if q, ok := p["question"].(string); ok {
				question = q
			} else if m, ok := p["message"].(string); ok {
				question = m
			}
			// Accept both []any (JSON-decoded payloads) and []string (Go code
			// that follows the documented `"options": []string{...}` pattern).
			switch opts := p["options"].(type) {
			case []any:
				options = opts
			case []string:
				options = make([]any, len(opts))
				for i, s := range opts {
					options[i] = s
				}
			}
		}
		actions := make([]MessageAction, 0, len(options))
		for _, o := range options {
			switch a := o.(type) {
			case string:
				actions = append(actions, MessageAction{Label: a})
			case MessageAction:
				actions = append(actions, a)
			case map[string]any:
				label, _ := a["label"].(string)
				value, _ := a["value"].(string)
				actions = append(actions, MessageAction{Label: label, Value: value})
			}
		}
		return []FrameMapping{{textFrame(question, Frame{"actions": actions}), true}}
	case "busy":
		return []FrameMapping{{textFrame("⏳ Still working on the previous message — one moment.", nil), false}}
	case "run_started":
		return []FrameMapping{{Frame{"type": "run", "data": map[string]any{"status": "started"}}, false}}
	case "run_finished":
		return []FrameMapping{{Frame{"type": "run", "data": map[string]any{"status": "finished"}}, false}}
	case "run_error":
		return []FrameMapping{
			{textFrame("⚠️ "+ev.Error, nil), true},
			{Frame{"type": "run", "data": map[string]any{"status": "finished"}}, false},
		}
	}
	return nil
}

File: packages/core/dotnet/Protocol.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Botiva;

/// <summary>Parsed user message.</summary>
public sealed record IncomingMessage(string Text, string? Id = null, JsonObject? Meta = null);

/// <summary>Client handshake.</summary>
public sealed record HelloFrame(string? UserId, string? ConversationId, int? Watermark, JsonObject? Meta);

/// <summary>ParseIncoming result — exactly one property is non-null.</summary>
public sealed record Inbound(HelloFrame? Hello, IncomingMessage? Message);

public sealed record FrameMapping(Frame Frame, bool Persistent);

public static class Protocol
{
    public const string Version = "botiva/1";

    private static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);

    public static JsonNode? ToNode(object? value) =>
        value is null ? null : JsonSerializer.SerializeToNode(value, Json);

    /// <summary>
    /// Accepts a JSON string, a parsed JsonObject or plain text — mirrors
    /// parseIncoming in @botiva/core.
    /// </summary>
    public static Inbound? ParseIncoming(object? raw)
    {
        JsonObject? value = null;
        switch (raw)
        {
            case string s:
                try
                {
                    value = JsonNode.Parse(s) as JsonObject;
                }
                catch (JsonException)
                {
                    var text = s.Trim();
                    return text.Length == 0 ? null : new Inbound(null, new IncomingMessage(text));
                }
                break;
            case JsonObject o:
                value = o;
                break;
        }
        if (value is null) return null;

        if ((string?)value["type"] == "hello")
        {
            return new Inbound(new HelloFrame(
                (string?)value["userId"],
                (string?)value["conversationId"],
                value["watermark"] is JsonNode w ? (int?)w.GetValue<double>() : null,
                value["meta"] as JsonObject), null);
        }

        var body = (string?)(value["data"]?["text"]) ?? (string?)value["text"] ?? "";
        body = body.Trim();
        if (body.Length == 0) return null;
        return new Inbound(null, new IncomingMessage(body, (string?)value["id"], value["meta"] as JsonObject));
    }

    /// <summary>
    /// Canonical AgentEvent → wire frame mapping — must stay byte-compatible
    /// with @botiva/core eventToFrames (PROTOCOL.md §4).
    /// </summary>
    public static List<FrameMapping> EventToFrames(AgentEvent ev, Func<string, string> newId)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        Frame Text(string text, Action<Frame>? extra = null)
        {
            var frame = new Frame
            {
                ["type"] = "text",
                ["id"] = newId("msg"),
                ["from"] = "bot",
                ["data"] = new JsonObject { ["text"] = text },
                ["timestamp"] = now,
            };
            extra?.Invoke(frame);
            return frame;
        }

        switch (ev.Type)
        {
            case "message":
                return [new(Text(ev.Text ?? "", f =>
                {
                    if (ev.Actions is { Count: > 0 }) f["actions"] = ToNode(ev.Actions);
                }), true)];
            case "tool_call":
                return [new(new Frame { ["type"] = "tool_call", ["data"] = ToNode(ev.ToolCall) }, true)];
            case "genui":
                return [new(new Frame
                {
                    ["type"] = "genui",
                    ["streamId"] = ev.StreamId ?? newId("stream"),
                    ["chunk"] = ToNode(ev.Chunk),
                    ["done"] = ev.Done,
                }, true)];
            case "interrupt":
            {
                var question = "Your confirmation is needed to continue.";
                JsonArray options = ["Approve", "Cancel"];
                var payload = ToNode(ev.Payload);
                if (payload is JsonValue v && v.TryGetValue<string>(out var s)) question = s;
                else if (payload is JsonObject p)
                {
                    question = (string?)p["question"] ?? (string?)p["message"] ?? question;
                    if (p["options"] is JsonArray arr) options = (JsonArray)arr.DeepClone();
                }
                var actions = new JsonArray();
                foreach (var option in options)
                {
                    actions.Add(option is JsonValue sv && sv.TryGetValue<string>(out var label)
                        ? new JsonObject { ["label"] = label }
                        : option?.DeepClone());
                }
                return [new(Text(question, f => f["actions"] = actions), true)];
            }
            case "busy":
                return [new(Text("⏳ Still working on the previous message — one moment."), false)];
            case "run_started":
                return [new(new Frame { ["type"] = "run", ["data"] = new JsonObject { ["status"] = "started" } }, false)];
            case "run_finished":
                return [new(new Frame { ["type"] = "run", ["data"] = new JsonObject { ["status"] = "finished" } }, false)];
            case "run_error":
                return
                [
                    new(Text($"⚠️ {ev.Error}"), true),
                    new(new Frame { ["type"] = "run", ["data"] = new JsonObject { ["status"] = "finished" } }, false),
                ];
            default:
                return [];
        }
    }
}
